chapter06/src/main/java/com/seaofnodes/simple/node/ScopeNode.java

```diff
@@ -126,13 +126,16 @@ public ScopeNode dup() {
      * @param that The ScopeNode to be merged into this
      * @return A new Region node representing the merge point
      */
-    public RegionNode mergeScopes(ScopeNode that) {
-        RegionNode r = (RegionNode)ctrl(new RegionNode(null,ctrl(), that.ctrl()).peephole());
-        for( int i=1; i<nIns(); i++ )
-            if( in(i) != that.in(i) ) // No need for redundant Phis
-                set_def(i,new PhiNode(_rlabels.get(i), r, in(i), that.in(i)).peephole());
-        that.kill();            // Kill merged scope
-        return r;
+    public Node mergeScopes(ScopeNode that) {
+        Node c = ctrl(new RegionNode(null,ctrl(), that.ctrl()).peephole());
+        if (c instanceof RegionNode r) {
```

Comment:
This seems wrong to me. It could be that the this scope was removed and needs to copy all variable state from `that`. Furthermore, `that` needs to be killed in any case.

Reply:
Yes agreed